### PR TITLE
tftp: T4012: Add TFTP VRF support

### PR DESCRIPTION
**Pull request by DmitriyEshenko**
## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes)
- [ ] Migration from an old Vyatta component to vyos-1x, please link to related PR inside obsoleted component
- [ ] Other (please describe):

## Related Task(s)

* https://phabricator.vyos.net/T4012

## Checklist:

- [X] I have read the [**CONTRIBUTING**](https://github.com/vyos/vyos-1x/blob/current/CONTRIBUTING.md) document
- [X] I have linked this PR to one or more Phabricator Task(s)
- [ ] I have run the components [**SMOKETESTS**](https://github.com/vyos/vyos-1x/tree/current/smoketest/scripts/cli) if applicable
- [X] My commit headlines contain a valid Task id
- [X] My change requires a change to the documentation
- [X] I have updated the documentation accordingly

### CODE REVIEW

File: src/conf_mode/tftp_server.py [hunk 1]
```diff
@@ -91,6 +93,9 @@ def generate(tftpd):
         else:
             config['listen_address'] = f'[{address}]:{port} -6'
 
+        if 'vrf' in tftpd['listen_address'][address]:
+            config['vrf'] = tftpd['listen_address'][address]['vrf']
```
Comment (c-po) [inline src/conf_mode/tftp_server.py]: Please change iterator to iterate over the dictionary using key/value pairs - `for address, address_config in tftp['listen_address'].items()`  to make reading the code easier and not iterate over multiple dictionaries.

File: data/templates/tftp-server/default.tmpl [hunk 2]
```diff
@@ -1,2 +1,7 @@
 ### Autogenerated by tftp_server.py ###
 DAEMON_ARGS="--listen --user tftp --address {{ listen_address }} {{ "--create --umask 000" if allow_upload is defined }} --secure {{ directory }}"
+{% if vrf is defined  %}
```
Comment (c-po) [inline data/templates/tftp-server/default.tmpl]: Please remove additional whitespace.

File: interface-definitions/include/listen-address-vrf.xml.i [hunk 3]
```diff
@@ -0,0 +1,25 @@
+<!-- include start from listen-address.xml.i -->
```
Comment (c-po) [inline interface-definitions/include/listen-address-vrf.xml.i]: Please fix include comment. References the wrong file.

File: src/conf_mode/tftp_server.py [hunk 4]
```diff
@@ -69,6 +70,7 @@ def verify(tftpd):
         if not is_addr_assigned(address):
             print(f'WARNING: TFTP server listen address "{address}" not ' \
                   'assigned to any interface!')
+        verify_vrf(tftpd['listen_address'][address])
```
Comment (c-po) [inline src/conf_mode/tftp_server.py]: Please change iterator to iterate over the dictionary using key/value pairs - `for address, address_config in tftp['listen_address'].items()` to make reading the code easier and not iterate over multiple dictionaries.

**Review by c-po — CHANGES_REQUESTED** [review]

File: src/conf_mode/tftp_server.py (hunk 1 above)
Comment (DmitriyEshenko) [reply src/conf_mode/tftp_server.py]: Done

File: data/templates/tftp-server/default.tmpl (hunk 2 above)
Comment (DmitriyEshenko) [reply data/templates/tftp-server/default.tmpl]: Done

File: interface-definitions/include/listen-address-vrf.xml.i (hunk 3 above)
Comment (DmitriyEshenko) [reply interface-definitions/include/listen-address-vrf.xml.i]: Done

File: src/conf_mode/tftp_server.py (hunk 4 above)
Comment (DmitriyEshenko) [reply src/conf_mode/tftp_server.py]: Done

File: src/systemd/tftpd@.service [hunk 5]
```diff
@@ -7,7 +7,7 @@ RequiresMountsFor=/run
 Type=forking
 #NotifyAccess=main
 EnvironmentFile=-/etc/default/tftpd%I
-ExecStart=/usr/sbin/in.tftpd "$DAEMON_ARGS"
+ExecStart=/bin/sh -c "${VRF_ARGS} /usr/sbin/in.tftpd ${DAEMON_ARGS}"
```
Comment (c-po) [inline src/systemd/tftpd@.service:10]: Is there a particular reason this uses `/bin/sh` as executor and not like the other patterns? Just trying to understand.

* https://github.com/vyos/vyos-1x/blob/current/data/templates/ssh/override.conf.tmpl#L9
* https://github.com/vyos/vyos-1x/blob/current/data/templates/ntp/override.conf.tmpl#L9

Comment (DmitriyEshenko) [reply src/systemd/tftpd@.service:10]: Yes, we can not use ENV arg as the first param if it is not defined exactly in service. In this case, it is not possible to define VRF_ARGS here and we are defining it in /etc/default/tftpN files.

**Comment by c-po** [comment]
Can you please also extend https://github.com/vyos/vyos-1x/blob/current/smoketest/scripts/cli/test_service_tftp-server.py to start some instances in different VRFs and then verify if the VRF runs a tftpd process (`ip vrf pids red`).

I see no documentation PR, but the template mentions documentation needs update?

**Pull request merged**